libs/zpool/src/handle.zig

```diff
@@ -57,78 +57,71 @@ pub fn Handle(
         64 => u64,
         128 => u128,
         256 => u256,
-        else => @compileError("index_bits + cycle_bits should sum to exactly " ++
+        else => @compileError("index_bits + cycle_bits must sum to exactly " ++
             "8, 16, 32, 64, 128, or 256 bits"),
     };
 
     const utils = @import("utils.zig");
     const UInt = utils.UInt;
     const AddressableUInt = utils.AddressableUInt;
 
-
     return struct {
         const Self = @This();
 
-        pub const Resource = TResource;
-        pub const Index = UInt(index_bits);
-        pub const Cycle = UInt(cycle_bits);
-
-        const Union = extern union {
+        const CompactHandle = Self;
+        const CompactIndex = UInt(index_bits);
+        const CompactCycle = UInt(cycle_bits);
+        const CompactUnion = extern union {
             id: Id,
-            handle: packed struct { index: Index, cycle: Cycle },
+            bits: packed struct {
+                index: CompactIndex,
+                cycle: CompactCycle,
+            },
         };
 
+        pub const Resource = TResource;
+
         pub const AddressableIndex = AddressableUInt(index_bits);
         pub const AddressableCycle = AddressableUInt(cycle_bits);
 
-        pub const max_index = ~@as(Index, 0);
-        pub const max_cycle = ~@as(Cycle, 0);
+        pub const max_index = ~@as(CompactIndex, 0);
+        pub const max_cycle = ~@as(CompactCycle, 0);
         pub const max_count = @as(Id, max_index - 1) + 2;
 
         id: Id = 0,
 
         pub const nil = Self{ .id = 0 };
 
-        pub fn init(_index: Index, _cycle: Cycle) Self {
-            var u = Union{ .handle = .{ .index = _index, .cycle = _cycle }};
+        pub fn init(index: CompactIndex, cycle: CompactCycle) Self {
+            var u = CompactUnion{ .bits = .{ .index = index, .cycle = cycle } };
             return .{ .id = u.id };
         }
 
+        /// Unpacks the `index` and `cycle` bit fields that comprise
+        /// `Handle.id` into an `AddressableHandle`, which stores
+        /// the `index` and `cycle` values in pointer-addressable fields.
         pub fn addressable(self: Self) AddressableHandle {
+            var u = CompactUnion{ .id = self.id };
             return .{
-                .index = self.index(),
-                .cycle = self.cycle(),
+                .index = u.bits.index,
+                .cycle = u.bits.cycle,
             };
         }
 
-        pub fn index(self: Self) AddressableIndex {
```

Comment:
Favor operating on `AddressableHandle` in all cases where `index` and `cycle` are needed.  The goal is an API that discourages shifting and masking the `index` and `cycle` bits for every operation.